javaServices/coreJavaServices/src/main/java/joliex/db/DatabaseService.java

```diff
@@ -523,24 +505,121 @@ static FaultException createFaultException( SQLException e ) {
+					.getPreparedStatement() ) {
+				resultValue = _executeQuery( stm, request );
+			} catch( SQLException e ) {
+				throw createFaultException( e );
+			}
+		} else {
+			int txHandle = request.getFirstChild( "txHandle" ).intValue();
+			Connection tx = _tryGetOpenTransaction( txHandle );
+			try( PreparedStatement stm = new NamedStatementParser( tx, request.getFirstChild( "query" ).strValue(),
+				request.getFirstChild( "query" ) )
+					.getPreparedStatement(); ) {
+				resultValue = _executeQuery( stm, request.getFirstChild( "query" ) );
+				openTxs.put( txHandle, tx );
 			} catch( SQLException e ) {
 				throw createFaultException( e );
 			}
 		}
+		return resultValue;
+	}
+
+	private FaultException createTransactionException( String message ) {
+		Value v = Value.create();
+		v.getNewChild( "message" ).setValue( message );
+		return new FaultException( "TransactionException", v );
+	}
 
+	private Value _executeQuery( PreparedStatement stm, Value request ) throws SQLException {
+		Value resultValue = Value.create();
+		ResultSet result = stm.executeQuery();
+		if( request.hasChildren( TEMPLATE_FIELD ) ) {
+			resultSetToValueVectorWithTemplate( result, resultValue.getChildren( "row" ),
+				request.getFirstChild( TEMPLATE_FIELD ) );
+		} else {
+			resultSetToValueVector( result, resultValue.getChildren( "row" ) );
+		}
+		result.close();
 		return resultValue;
 	}
+
+	@RequestResponse
+	public Value beginTx() throws FaultException {
+		_checkConnection();
+		Value response = Value.create();
+		Connection con;
+		try {
+			con = connectionPool.getConnection();
+			int txHandle = txHandles.getAndIncrement();
+			con.setAutoCommit( false );
+
+			openTxs.put( txHandle, con );
+
+			response.setValue( txHandle );
+			return response;
+		} catch( SQLException e ) {
+			throw createFaultException( e );
+		}
+	}
+
+	@RequestResponse
+	public void commitTx( Value request ) throws FaultException {
+		_checkConnection();
+
+		int txHandle = request.intValue();
+		Connection tx = openTxs.remove( txHandle );
+		try {
+			tx.commit();
+		} catch( SQLException e ) {
+			createFaultException( e );
+		} finally {
+			_tryCloseTransaction( tx );
+		}
+	}
+
+	@RequestResponse
+	public void rollbackTx( Value request ) throws FaultException {
+		_checkConnection();
+
+		int txHandle = request.intValue();
+		Connection tx = openTxs.remove( txHandle );
+		_tryCloseTransaction( tx );
+	}
+
+	private void _tryCloseTransaction( Connection con ) throws FaultException {
+		try {
+			con.setAutoCommit( true );
+			con.rollback();
+			con.close();
+		} catch( SQLException e ) {
+			try {
+				con.close();
+			} catch( SQLException e1 ) {
+				throw createFaultException( e );
+			}
+		}
+	}
+
+	private HikariDataSource _createDataSource() {
+		HikariConfig config = new HikariConfig();
+
+		config.setUsername( username );
+		config.setPassword( password );
+		config.setMaximumPoolSize( 6 );
```

Comment:
I am very uncertain what would be a good pool-size, so for now it is hard-coded to 6.

[Microsoft sql](https://learn.microsoft.com/en-us/dotnet/framework/data/adonet/sql-server-connection-pooling) sets a connection pool size of max 100.

Connections are by default automatically added to the pool if needed, and removed after having been idle for 10 minutes, so maybe increasing this number is not bad

Reply:
Should probably be a configuration parameter.